pip/pip-415.md

```diff
@@ -0,0 +1,186 @@
+| **403 Forbidden**           | Client lacks permissions to access the topic/namespace.                                                                                                      |       
+| **404 Not Found**           | Topic/namespace does not exist, or the specified `index` is invalid. An index value of -1 or one that exceeds the maximum index range is considered invalid. |
+| **406 Not Acceptable**      | The topic is not a persistent topic.                                                                                                                         |
+| **412 Precondition Failed** | The broker is not configured to enable broker entry metadata.                                                                                                |
+
+#### Pulsar Admin
+Add two api in `pulsar-client-admin-api/src/main/java/org/apache/pulsar/client/admin/Topics.java`
+```java
+    /**
+     * Get the message id by index.
+     * @param topicName either the specific partition name of a partitioned topic (e.g. my-topic-partition-0) or the original topic name for non-partitioned topics.
+     * @param index the index of a message
+     * @return the message id of the message. 
+     * When retrieving a message ID by index, the resolution is limited to the **entry** level (an entry is the minimal storage unit for messages in Pulsar's persistence layer).
+     * If message batching is enabled, a single entry may contain multiple messages with distinct indexes.
+     * Example Scenario (partition with 2 entries):
+     * | Entry | Ledger ID | Entry ID | Index | Messages |
+     * | :--- | ---: | ---: | ---: | ---: |
+     * | A | 0 | 0 | 2 | 0,1,2 |
+     * | B | 0 | 1 | 4 | 3,4 |
+     * Param with indexes 0,1,2 or 3,4 will return the **same MessageID** (e.g., `MessageId(0:0:*)` for Entry A).
+     */
+    MessageId getMessageIdByIndex(String topicName, long index) throws PulsarAdminException;
+
+    /**
+     * Get the message id by index asynchronously.
+     * @param topicName either the specific partition name of a partitioned topic (e.g. my-topic-partition-0) or the original topic name for non-partitioned topics.
+     * @param index the index of a message
+     * @return the message id of the message
+     * When retrieving a message ID by index, the resolution is limited to the **entry** level (an entry is the minimal storage unit for messages in Pulsar's persistence layer).
+     * If message batching is enabled, a single entry may contain multiple messages with distinct indexes.
+     * Example Scenario (partition with 2 entries):
+     * | Entry | Ledger ID | Entry ID | Index | Messages |
+     * | :--- | ---: | ---: | ---: | ---: |
+     * | A | 0 | 0 | 2 | 0,1,2 |
+     * | B | 0 | 1 | 4 | 3,4 |
+     * Param with indexes 0,1,2 or 3,4 will return the **same MessageID** (e.g., `MessageId(0:0:*)` for Entry A).
+     */
+    CompletableFuture<MessageId> getMessageIdByIndexAsync(String topicName, long index);
```

Comment:
We need to add a `@throws` here to indicate that when the index is invalid, the future will complete exceptionally with `PulsarAdminException#NotFoundException`.

Reply:
According to convention, we typically do not include @throws in Javadoc comments for asynchronous methods. Perhaps using @implNote to clarify could be an alternative approach.